blockdevice/encryption/luks/luks.go

```diff
@@ -0,0 +1,238 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package luks provides a way to call LUKS2 cryptsetup.
+package luks
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/talos-systems/go-blockdevice/blockdevice/encryption"
+	"github.com/talos-systems/go-blockdevice/blockdevice/filesystem/luks"
+	"github.com/talos-systems/go-blockdevice/blockdevice/util"
+)
+
+// Cipher LUKS2 cipher type.
+type Cipher int
+
+// String converts to command line string parameter value.
+func (c Cipher) String() (string, error) {
+	switch c {
+	case AESXTSPlain64Cipher:
+		return AESXTSPlain64CipherString, nil
+	default:
+		return "", fmt.Errorf("unknown cipher kind %d", c)
+	}
+}
+
+// ParseCipherKind converts cipher string into cipher type.
+func ParseCipherKind(s string) (Cipher, error) {
+	switch s {
+	case "": // default
+		fallthrough
+	case AESXTSPlain64CipherString:
+		return AESXTSPlain64Cipher, nil
+	default:
+		return 0, fmt.Errorf("unknown cipher kind %s", s)
+	}
+}
+
+const (
+	// AESXTSPlain64CipherString string representation of aes-xts-plain64 cipher.
+	AESXTSPlain64CipherString = "aes-xts-plain64"
+	// AESXTSPlain64Cipher represents aes-xts-plain64 encryption cipher.
+	AESXTSPlain64Cipher Cipher = iota
+)
+
+// LUKS implements LUKS2 encryption provider.
+type LUKS struct {
+	cipher                 Cipher
+	argonBenchmarkDeadline time.Duration
+	argonIterations        uint
+	argonMemoryLimit       uint64
+}
+
+// New creates new LUKS2 encryption provider.
+func New(cipher Cipher, argonBenchmarkDeadline time.Duration, argonIterations uint, argonMemoryLimit uint64) *LUKS {
+	return &LUKS{
+		cipher:                 cipher,
+		argonBenchmarkDeadline: argonBenchmarkDeadline,
+		argonIterations:        argonIterations,
+		argonMemoryLimit:       argonMemoryLimit,
+	}
+}
+
+// Open runs luksOpen on a device and returns mapped device path.
+func (l *LUKS) Open(deviceName string, password []byte) (string, error) {
+	parts := strings.Split(deviceName, "/")
+	mappedPath := util.PartPathEncrypted(parts[len(parts)-1])
+	parts = strings.Split(mappedPath, "/")
+	mappedName := parts[len(parts)-1]
+
+	err := l.runCommand([]string{"luksOpen", deviceName, mappedName, "--key-file=-"}, password)
+	if err != nil {
+		return "", err
+	}
+
+	return mappedPath, nil
+}
+
+// Encrypt implements encryption.Provider.
+func (l *LUKS) Encrypt(deviceName string, passwords ...interface{}) error {
```

Comment:
I feel using loosely-typed arguments here might be not the best solution, as it avoids compile-time checks.

Can we do instead `[]Passphrase`, where `struct Passphrase { slot int; password []byte} `? 

And same type `Passphrase` can be used in other functions